### PR TITLE
CIRC-7337: Add metric name parsing to gosnowth

### PR DESCRIPTION
**Pull request by dhaifley**
* add: Added metric name parsing functionality via the ParseMetricName() function and its associated types.

### CODE REVIEW

File: metric_test.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+			num:   3,
+			lit:   "testing",
+		},
+		{
+			input: `testing|ST[b"QUFB":blah,blah1:b"QkJCQgo=",b"YWFh":b"YmJi"]`,
+			num:   3,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah:blah:blah]",
```
Comment (bdunavant) [inline metric_test.go:178]:
```suggestion
			input: "testing|ST[blah:blah:blah:blah]",
```
This test isn't actually valid as this is not a valid tag format.

File: metric_test.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+			num:   1,
+			lit:   "testing",
+		},
```
Comment (bdunavant) [inline metric_test.go:181]: The test suite for this only tests the metric names themselves, and not that the tag parsing is also working correctly.   There are also no tests for measurement tags, or names in a valid, but non-canonical form.  We also now have support for quoted strings in tags.   We likely want to add these to the test suite.   examples:

`foo|ST[a:b]|ST[b:c]` multiple `|ST[]` clauses
`foo|MT[a:b]` measurement tags
`foo|MT[a:b]|ST[b:d,e:f]|MT[g:h]|ST[i:j]` multiples of both
`foo|ST["quoted key string":"quoted value string"]`

https://docs.circonus.com/irondb/metric-names-tags-queries/ goes into more detail on what is valid or not.

File: metric_test.go (hunk 1 above)
Comment (dhaifley) [reply metric_test.go:188]: This has been corrected.

File: metric_test.go (hunk 2 above)
Comment (dhaifley) [reply metric_test.go:234]: There was not yet support for measurement tags or multiple tag sets or quoted key and value strings in this parser. All of this functionality has now been added, and these unit tests have been added to the test suite.

File: metric_test.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+		{
+			input: `testing|ST["blah:|ST[]":blah]|MT[blah:",]:|MTblah"]`,
+			numST: 1,
+			numMT: 1,
+			lit:   "testing",
+		},
+		{
+			input: `testing|ST["b:|ST[]":b]|MT[b:",]:|MTb"]|ST[a:b]|MT[c:d]`,
```
Comment (bdunavant) [inline metric_test.go:230]: Please add a test for `testing|ST["quote\"slash\\":bar]` which is has tag `quote"slash\` with value 'bar'.

File: metric.go [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+	case rune(0):
+		return tokenEOF, ""
+	case '[':
+		return tokenOB, string(ch)
+	case ']':
+		return tokenCB, string(ch)
+	case '"':
+		return tokenQuote, string(ch)
+	case 'b':
+		return tokenB64, string(ch)
```
Comment (pheckenlWork) [inline metric.go:101]: I noticed in package tests, tokenQuote and tokenB64 are not getting exercised.  Is that expected?

**Review by pheckenlWork — COMMENTED** [review]
Since this is a parser, have you thought about throwing the beta of go fuzz against it?  Go fuzz is supposed to move into the standard library in Go 18.

Here is a link: https://go.dev/blog/fuzz-beta

File: metric_test.go (hunk 3 above)
Comment (dhaifley) [reply metric_test.go:230]: I have now added support for escaped characters within quoted text in the metric name. I have added this test to the test suite as well.

File: metric.go (hunk 4 above)
Comment (dhaifley) [reply metric.go]: Good catch. The linter showed me that these are actually not being used in the parser, there is no case where scan() is used that these would be found. I have removed these cases.

**Comment by dhaifley** [comment]
I agree this would be a great candidate for fuzz testing. I think I am going to wait on actually committing any fuzz tests to this repo tho until go 1.18 is out and we have tested it and moved to using it internally.

**Review by bdunavant — APPROVED** [review]
LGTM

**Pull request merged**